Plugin Update Workflow Does not Remove Old Plugin And Causes App To Not Start

When a plugin is updated through SQ the old plugin file does not get removed on the restart and causes startup errors due to 2 versions of a plugin being present.

Additional call stack that is related.
2019.11.07 05:03:18 ERROR web[][o.s.s.p.PlatformImpl] Web server startup failed
java.lang.IllegalStateException: Unable to read plugin manifest from jar : D:\home\site\wwwroot\sonarqube-8.0\extensions\plugins\sonar-vbnet-plugin-8.0.0.9566.jar
	at org.sonar.updatecenter.common.PluginManifest.<init>(PluginManifest.java:125)
	at org.sonar.core.platform.PluginInfo.create(PluginInfo.java:412)
	at org.sonar.server.plugins.ServerPluginRepository.overrideAndRegisterPlugin(ServerPluginRepository.java:189)
	at org.sonar.server.plugins.ServerPluginRepository.moveDownloadedPlugins(ServerPluginRepository.java:146)
	at org.sonar.server.plugins.ServerPluginRepository.start(ServerPluginRepository.java:104)
	at org.sonar.core.platform.StartableCloseableSafeLifecyleStrategy.start(StartableCloseableSafeLifecyleStrategy.java:40)
	at org.picocontainer.injectors.AbstractInjectionFactory$LifecycleAdapter.start(AbstractInjectionFactory.java:84)
	at org.picocontainer.behaviors.AbstractBehavior.start(AbstractBehavior.java:169)
	at org.picocontainer.behaviors.Stored$RealComponentLifecycle.start(Stored.java:132)
	at org.picocontainer.behaviors.Stored.start(Stored.java:110)
	at org.picocontainer.DefaultPicoContainer.potentiallyStartAdapter(DefaultPicoContainer.java:1016)
	at org.picocontainer.DefaultPicoContainer.startAdapters(DefaultPicoContainer.java:1009)
	at org.picocontainer.DefaultPicoContainer.start(DefaultPicoContainer.java:767)
	at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:135)
	at org.sonar.server.platform.platformlevel.PlatformLevel.start(PlatformLevel.java:90)
	at org.sonar.server.platform.platformlevel.PlatformLevel2.start(PlatformLevel2.java:109)
	at org.sonar.server.platform.PlatformImpl.start(PlatformImpl.java:213)
	at org.sonar.server.platform.PlatformImpl.startLevel2Container(PlatformImpl.java:179)
	at org.sonar.server.platform.PlatformImpl.init(PlatformImpl.java:87)
	at org.sonar.server.platform.web.PlatformServletContextListener.contextInitialized(PlatformServletContextListener.java:43)
	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4770)
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5236)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1423)
	at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1413)
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.nio.file.NoSuchFileException: D:\home\site\wwwroot\sonarqube-8.0\extensions\plugins\sonar-vbnet-plugin-8.0.0.9566.jar
	at java.base/sun.nio.fs.WindowsException.translateToIOException(Unknown Source)
	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(Unknown Source)
	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(Unknown Source)
	at java.base/sun.nio.fs.WindowsFileAttributeViews$Basic.readAttributes(Unknown Source)
	at java.base/sun.nio.fs.WindowsFileAttributeViews$Basic.readAttributes(Unknown Source)
	at java.base/sun.nio.fs.WindowsFileSystemProvider.readAttributes(Unknown Source)
	at java.base/java.nio.file.Files.readAttributes(Unknown Source)
	at java.base/java.util.zip.ZipFile$Source.get(Unknown Source)
	at java.base/java.util.zip.ZipFile$CleanableResource.<init>(Unknown Source)
	at java.base/java.util.zip.ZipFile$CleanableResource.get(Unknown Source)
	at java.base/java.util.zip.ZipFile.<init>(Unknown Source)
	at java.base/java.util.zip.ZipFile.<init>(Unknown Source)
	at java.base/java.util.jar.JarFile.<init>(Unknown Source)
	at java.base/java.util.jar.JarFile.<init>(Unknown Source)
	at java.base/java.util.jar.JarFile.<init>(Unknown Source)
	at org.sonar.updatecenter.common.PluginManifest.<init>(PluginManifest.java:120)
	... 28 common frames omitted

--- a/wwwroot/HttpPlatformHandlerStartup.ps1
+++ b/wwwroot/HttpPlatformHandlerStartup.ps1
@@ -90,6 +90,21 @@ log("Writing to wrapper.conf file")	log('Updating wrapper.conf based on environm
 $wrapperConfigContents = Get-Content -Path $wrapperConfig.FullName -Raw
 $wrapperConfigContents -ireplace 'wrapper\.java\.command=.*', 'wrapper.java.command=%JAVA_HOME%\bin\java' | Set-Content -Path $wrapperConfig.FullName -NoNewLine
 
+log('Searching for duplicate plugins.')
+$plugins = Get-ChildItem '.\sonarqube-*\extensions\plugins\*' -Filter '*.jar'
+$pluginBaseName = $plugins | ForEach-Object { $_.Name.substring(0, $_.Name.LastIndexOf('-')) }
+$uniquePlugins = $pluginBaseName | Select-Object -Unique
+$duplicates = Compare-Object -ReferenceObject $uniquePlugins -DifferenceObject $pluginBaseName
+if($duplicates)
+{
+    log("Duplicates plugins found for: $($duplicates.InputObject)")a
+    foreach($duplicate in $duplicates) { 
+        $oldestFile = $plugins | Where-Object { $_.Name -imatch $duplicate.InputObject } | Sort-Object -Property 'LastWriteTime' | Select-Object -First 1
+        log("Deleting $oldestFile")
+        $oldestFile | Remove-Item
+    }
+}
+
 log('Searching for StartSonar.bat')
 $startScript = Get-ChildItem 'StartSonar.bat' -Recurse
 if(!$startScript) {

